### PR TITLE
Upgrading to Nuke 6

### PR DESCRIPTION
**Pull request by slewis74**

### CODE REVIEW

File: source/NukeBuildComponents/IOctopusNukeBuild.cs [hunk 1]
```diff
@@ -13,9 +12,9 @@ public interface IOctopusNukeBuild : INukeBuild
 
         Enumeration Config { get; }
 
-        [Solution] Solution Solution => ValueInjectionUtility.TryGetValue(() => Solution);
+        [Solution] Solution? Solution => TryGetValue(() => Solution);
 
-        [OctoVersion] OctoVersionInfo OctoVersionInfo => ValueInjectionUtility.TryGetValue(() => OctoVersionInfo);
+        [OctoVersion] OctoVersionInfo? OctoVersionInfo => TryGetValue(() => OctoVersionInfo);
```
Comment (cailyoung) [inline source/NukeBuildComponents/IOctopusNukeBuild.cs:17]: I think that making this nullable might be causing `IPackComponent` to set empty string outputs for `semver` and `prerelease_tag`, which is causing GHA to fail on the tag action (and that seems to have pushed a `v1.0.0` for some reason...)

Comment (cailyoung) [reply source/NukeBuildComponents/IOctopusNukeBuild.cs:17]: Yeah, see [here](https://github.com/OctopusDeploy/NukeBuildComponents/runs/5843967911?check_suite_focus=true#step:4:139) - `null` in the output.

**Review by cailyoung — CHANGES_REQUESTED** [review]
Hey there! Thanks for the contribution. Please consider putting some context in the PR description as to why the bump is happening so that we can consider impacts.

You may want to also include a bump to the .NET version specified in the workflow file for `setup-dotnet`; I _think_ Nuke can deal with it, but if we're going to bother with `setup-dotnet` let's have the versions match.

File: source/NukeBuildComponents/ICompile.cs [hunk 2]
```diff
@@ -12,12 +13,12 @@ public interface ICompile : IOctopusNukeBuild
...
             .Executes(() =>
             {
-                Logger.Info("Building {1} v{0}", OctoVersionInfo.FullSemVer, TargetPackageDescription);
+                Log.Information("Building {1} v{0}", OctoVersionInfo?.FullSemVer, TargetPackageDescription);
 
                 DotNetBuild(_ => _
                     .SetProjectFile(Solution)
                     .SetConfiguration(Config)
-                    .SetVersion(OctoVersionInfo.FullSemVer)
+                    .SetVersion(OctoVersionInfo?.FullSemVer)
```
Comment (matt-richardson) [inline source/NukeBuildComponents/ICompile.cs:21]:
```suggestion
                    .SetVersion(OctoVersionInfo!.FullSemVer)
```

File: source/NukeBuildComponents/ICleanCode.cs [hunk 3]
```diff
@@ -3,28 +3,30 @@
...
+
         Target CleanCode => _ => _
             .OnlyWhenStatic(() => !IsLocalBuild && ShouldCleanCode)
             .TryTriggeredBy<IPackComponent>(x => x.Pack)
             .TryTriggeredBy<IPackExtension>(x => x.Pack)
             .Executes(() =>
             {
                 ReSharperTasks.ReSharperCleanupCode(new ReSharperCleanupCodeSettings()
-                    .SetTargetPath(Solution.Path));
+                    .SetTargetPath(Solution?.Path));
```
Comment (matt-richardson) [inline source/NukeBuildComponents/ICleanCode.cs:21]: Nearly all these should be `!` (can we fix it at source?)

File: source/NukeBuildComponents/IPackComponent.cs [hunk 4]
```diff
@@ -12,21 +13,21 @@ public interface IPackComponent : IOctopusNukeBuild
...
+                Log.Information("Packing {1} v{0}", OctoVersionInfo?.FullSemVer, TargetPackageDescription);
 
                 // This is done to pass the data to github actions
-                Console.Out.WriteLine($"::set-output name=semver::{OctoVersionInfo.FullSemVer}");
-                Console.Out.WriteLine($"::set-output name=prerelease_tag::{OctoVersionInfo.PreReleaseTagWithDash}");
+                Console.Out.WriteLine($"::set-output name=semver::{OctoVersionInfo?.FullSemVer}");
+                Console.Out.WriteLine($"::set-output name=prerelease_tag::{OctoVersionInfo?.PreReleaseTagWithDash}");
 
-                DotNetPack(_ => _
+                DotNetPack(c => c
```
Comment (matt-richardson) [inline source/NukeBuildComponents/IPackComponent.cs:22]: nuke convention is `_` here

File: source/NukeBuildComponents/IOctopusNukeBuild.cs [hunk 5]
```diff
@@ -13,13 +17,43 @@ public interface IOctopusNukeBuild : INukeBuild
...
+        OctoVersionInfo OctoVersionInfo => TryGetInjectedOctoVersionInfo(() => OctoVersionInfo);
 
         AbsolutePath SourceDirectory => RootDirectory / "source";
         public AbsolutePath ArtifactsDirectory => RootDirectory / "artifacts";
         public AbsolutePath PublishDirectory => RootDirectory / "publish";
         public AbsolutePath LocalPackagesDir => RootDirectory / ".." / "LocalPackages";
+
+        private static readonly Dictionary<MemberInfo, object> s_valueCache = new Dictionary<MemberInfo, object>();
+
+        // we need this custom method instead of the "normal" TryGetValue,
```
Comment (matt-richardson) [inline source/NukeBuildComponents/IOctopusNukeBuild.cs:44]: Important note here

File: source/NukeBuildComponents/OctoVersionThatWorksWithBuildComponentsAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            TeamCity.Instance?.SetBuildNumber(version.FullSemVer);
+            AppVeyor.Instance?.UpdateBuildVersion(version.FullSemVer);
+        }
+
+        return version;
+    }
+
+    protected new T GetMemberValue<T>(string memberName, object instance)
+    {
+        //Note: this is the only real change we need here:
```
Comment (matt-richardson) [inline source/NukeBuildComponents/OctoVersionThatWorksWithBuildComponentsAttribute.cs:171]: Important note here

**Review by matt-richardson — APPROVED** [review]
I think we can fix up those nullability things - we want to go :boom: if things are null instead of going 🙈 .

I'm not enthused about the horrible solution that I've contributed, but it gets us out of the hole for now.
I've reached out to Mattias (nuke maintainer) about it, he's going to look into a better fix at source. Once that happens we can revert this (I don't want this code to live on forever).

**Pull request merged**